File: src/main/java/com/intetics/lukyanenko/dao/jdbc/CustomerDAOImpl.java
package com.intetics.lukyanenko.dao.jdbc;

import com.intetics.lukyanenko.dao.CustomerDAO;
import com.intetics.lukyanenko.models.Customer;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;

public class CustomerDAOImpl
  extends CommonDAOImpl<Customer>
  implements CustomerDAO
{
  @Override
  protected String getBaseSelectSQL()
  {
    return "select * from customer";
  }
  
  @Override
  protected Customer mapFields(ResultSet resultSet, String fieldNamePrefix)
  throws SQLException
  {
    Customer object = getNewModelInstance();
    object.setId(resultSet.getInt(fieldNamePrefix + "id"));
//    object.setAnonymousSessionID(resultSet.getString("anonymous_session_id"));
    object.setAppUserName(resultSet.getString(fieldNamePrefix + "app_user_name"));
    object.setEmail(resultSet.getString(fieldNamePrefix + "email"));
    object.setPhone(resultSet.getString(fieldNamePrefix + "phone"));
    object.setShippingAddress(resultSet.getString(fieldNamePrefix + "shipping_address"));
    return object;
  }
  
  public void add(Customer model)
  {
    HashMap<String, Object> params = new HashMap<>();
    int id = generateNewID();
    params.put("id", id);
    params.put("app_user_name", model.getAppUserName());
    params.put("email", model.getEmail());
    params.put("phone", model.getPhone());
    params.put("shipping_address", model.getShippingAddress());
    params.put("anonymous_session_id", model.getAnonymousSessionID());
    jdbcTemplate.update("insert " +
                        "  into customer (" +
                        "    id," +
                        "    app_user_name," +
                        "    email," +
                        "    phone," +
                        "    shipping_address," +
                        "    anonymous_session_id" +
                        "  ) values (" +
                        "    :id," +
                        "    :app_user_name," +
                        "    :email," +
                        "    :phone," +
                        "    :shipping_address," +
                        "    :anonymous_session_id" +
                        "  )",
                        params
    );
    model.setId(id);
  }
  
  public void update(Customer model)
  {
    HashMap<String, Object> params = new HashMap<>();
    params.put("id", model.getId());
    params.put("app_user_name", model.getAppUserName());
    params.put("email", model.getEmail());
    params.put("phone", model.getPhone());
    params.put("shipping_address", model.getShippingAddress());
    params.put("anonymous_session_id", model.getAnonymousSessionID());
    jdbcTemplate.update("update customer (" +
                        "  set app_user_name    = :app_user_name," +
                        "      email            = :email," +
                        "      phone            = :phone," +
                        "      shipping_address = :shipping_address," +
                        "      anonymous_session_id = :anonymous_session_id" +
                        "  where id = :id",
                        params
    );
  }
  
  public void delete(Customer model)
  {
    HashMap<String, Object> params = new HashMap<>();
    params.put("id", model.getId());
    jdbcTemplate.update("delete from customer (" +
                        "  where id = :id" +
                        "    and not exists (select * " +
                        "                      from order_head" +
                        "                      where customer_id = :id)",
                        params
    );
  }
  
  public int generateNewID()
  {
    return jdbcTemplate.queryForObject("select next value for sq_customer",
                                       (HashMap<String, Object>)null,
                                       Integer.class);
  }
  
  @Override
  public Customer find(String sessionId)
  {
    return get("anonymous_session_id", sessionId);
  }
}

File: src/main/java/com/intetics/lukyanenko/dao/jdbc/CustomerDAOImpl.kt
package com.intetics.lukyanenko.dao.jdbc

import com.intetics.lukyanenko.dao.CustomerDAO
import com.intetics.lukyanenko.models.Customer
import org.springframework.stereotype.Service
import java.sql.ResultSet
import java.util.*

@Service
class CustomerDAOImpl : CommonDAOImpl<Customer>(), CustomerDAO {
    private val sqlInsert = """
        insert
          into customer (
            id,
            app_user_name,
            email,
            phone,
            shipping_address,
            anonymous_session_id
          ) values (
            :id,
            :app_user_name,
            :email,
            :phone,
            :shipping_address,
            :anonymous_session_id
          )""".trimIndent()
    private val sqlUpdate = """
        update customer (
          set app_user_name    = :app_user_name,
              email            = :email,
              phone            = :phone,
              shipping_address = :shipping_address,
              anonymous_session_id = :anonymous_session_id
          where id = :id""".trimIndent()
    private val sqlDelete = """
        delete from customer (
          where id = :id
            and not exists (select *
                              from order_head
                              where customer_id = :id
                           )""".trimIndent()

    override fun getBaseSelectSQL() = "select * from customer"

    override fun mapFields(resultSet: ResultSet, fieldNamePrefix: String): Customer {
        val obj = newModelInstance()
        obj.id = resultSet.getInt(fieldNamePrefix + "id")
        //    object.setAnonymousSessionID(resultSet.getString("anonymous_session_id"));
        obj.appUserName = resultSet.getString(fieldNamePrefix + "app_user_name")
        obj.email = resultSet.getString(fieldNamePrefix + "email")
        obj.phone = resultSet.getString(fieldNamePrefix + "phone")
        obj.shippingAddress = resultSet.getString(fieldNamePrefix + "shipping_address")
        return obj
    }

    override fun add(model: Customer) {
        val params = HashMap<String, Any?>()
        val id = generateNewID()
        params.put("id", id)
        params.put("app_user_name", model.appUserName)
        params.put("email", model.email)
        params.put("phone", model.phone)
        params.put("shipping_address", model.shippingAddress)
        params.put("anonymous_session_id", model.anonymousSessionID)
        jdbcTemplate.update(sqlInsert, params)
        model.id = id
    }

    override fun update(model: Customer) {
        val params = HashMap<String, Any?>()
        params.put("id", model.id)
        params.put("app_user_name", model.appUserName)
        params.put("email", model.email)
        params.put("phone", model.phone)
        params.put("shipping_address", model.shippingAddress)
        params.put("anonymous_session_id", model.anonymousSessionID)
        jdbcTemplate.update(sqlUpdate, params)
    }

    override fun delete(model: Customer) {
        jdbcTemplate.update(sqlDelete, mapOf("id" to model.id))
    }

    override fun generateNewID(): Int {
        return jdbcTemplate.queryForObject("select next value for sq_customer",
                mapOf<String, Any>(),
                Int::class.java)
    }

    override fun find(sessionId: String): Customer? {
        return get("anonymous_session_id", sessionId)
    }
}
